sklearn/neighbors/__init__.py

```diff
@@ -3,8 +3,8 @@
 algorithm.
 """
 
-from ._ball_tree import BallTree
-from ._kd_tree import KDTree
+from ._ball_tree import BallTree64 as BallTree
+from ._kd_tree import KDTree64 as KDTree
```

Comment:
At the moment we have:

```python
In [1]: import numpy as np

In [2]: from sklearn.neighbors import KDTree

In [3]: X = np.random.rand(100, 5)

In [4]: t = KDTree(X)

In [5]: t
Out[5]: <sklearn.neighbors._kd_tree.KDTree64 at 0x55ba08b84ac0>
```

We might want to have `KDTree` and `BallTree` (as Python interfaces) keep their UX and contract unchanged and hide thei type-specific implementation (using something _à la_ [type erasure](https://www.modernescpp.com/index.php/type-erasure)).

I think this can be done via composition, e.g. `KDTree` can remain the actual public python class callers are using and have a `KDTree{32,64}` instance as a private attribute (the implementation is inferable from `X.dtype` given by `KDTree`'s constructor). Calls on `KDTree`'s methods could then be dispatched to this component's methods.

Apart from improving the design, I think this approach will prevent breakage in downstream libraries that are using `isinstance(object, KDTree)`.

Everything expressed here for `KDTree` also applies to `BallTree`.

@Micky774: we have not chosen this design for `DistanceMetrics`, but we probably should have. What do you think?

Reply:
Hmm that is indeed a nicer solution. So, in the case of `DistanceMetric` it would look something like:

```python
dst = DistanceMetric(metric='manhattan', dtype=np.float32)

# Actually calls ManhattanDistance32.dist()
out = dst.dist()
```

If that is the case I agree it is worth including in this PR. It is less invasive/complex than the other approaches considered so far and leads to no change to the UX.

Reply:
Thank you. I agree that sounds like a really nice solution. I'll try it out.

Reply:
@jjerphan How can I declare the type of the private attribute because we are inferring the specific class KDTree64 or KDTree32 in the __init__ method. If I don't declare the type however I get this error AttributeError: 'sklearn.neighbors._ball_tree.BallTree' object has no attribute '__ball_tree'

Reply:
@OmarManzoor: I am trying to infer the situation you are facing. If `sklearn.neighbors._ball_tree.BallTree` is a Python class (with no adherence to Cython) which has a `__ball_tree` attribute (the actual `BallTree{32,64}` instance implemented as a Cython extension type), then no type has to be specified for `__ball_tree` which is seen as a generic Python object.

Can you point to a commit (which you can push on a temporary branch if needed) where you meet this problem?

Reply:
However now I am facing an issue where X can be a list and we can't infer the dtype. Kindly check this test for an example of this case
https://github.com/scikit-learn/scikit-learn/pull/25914/files#diff-32025910a5cea60474862d3699a076ac08a51b4eee1e0f48815a03a0f32299e1R54-R67

Reply:
It seems like this is not so straightforward. In various places in the code some attributes like data are actually accessed in the python code. That would mean with this structure we would need to set the _ball_tree or _kd_tree attribute as public and change all places which use .data to ._ball_tree.data